server/docker_resource.py

```diff
@@ -0,0 +1,433 @@
+
+
+import six
+import json
+from six import iteritems
+import hashlib
+
+from girder.api.v1.resource import Resource, RestException
+from .constants import PluginSettings
+from girder.utility.model_importer import ModelImporter
+from girder.api.rest import getCurrentUser
+from girder.api import access
+from girder.api.describe import Description, describeRoute
+
+
+class DockerResource(Resource):
+    """
+    Manages the exposed rest api. When the settings are updated te new list
+    of docker images is checked, pre-loaded images will be ignored. New images
+    will cause a job to generate the cli handler and generate the rest endpoint
+    asynchronously.Deleted images will result in the removal of the rest api
+    endpoint though docker will still cache the image unless removed manually
+    (docker rmi image_name)
+    """
+    resourceName = ''
+
+    def __init__(self, name):
+        super(DockerResource, self).__init__()
+        self.resourceName = name
+        DockerResource.resourceName = name
+        self.route('PUT', (DockerResource.resourceName, 'docker_image'),
+                   self.appendImage)
+        self.route('DELETE', (DockerResource.resourceName, 'docker_image'),
+                   self.deleteImage)
+        self.route('GET', (DockerResource.resourceName, 'docker_image'),
+                   self.getDockerImages)
+
+    @access.admin
+    @describeRoute(
+        Description('list docker images and their clis ')
+        .notes("""Must be a system administrator to call this. If the value
+                       passed is a valid JSON object.""")
+        .errorResponse(
+            'You are not a system administrator.', 403)
+        .errorResponse(
+            'Failed to set system setting.', 500)
+    )
+    def getDockerImages(self, params):
+        data = {}
+        current = DockerCache(getDockerImageSettings())
+        for image in current.getDockerImageList():
+            data[image] = current.getCLIDict(image)
+        return data
+
+    @access.admin
+    @describeRoute(
+        Description('Remove a docker image ')
+        .notes(
+            """Must be a system administrator to call this. If the value
+                       passed is a valid JSON object.""")
+        .param('name', 'The name or a list of names  of the '
+                       'docker images to be removed', required=True)
+        .errorResponse('You are not a system administrator.', 403)
+        .errorResponse('Failed to set system setting.', 500)
+    )
+    # TODO delete REST endpoint
+    def deleteImage(self, params):
+
+        self.requireParams(('name',), params)
+        name = params['name']
+        name = json.loads(name)
+
+        if isinstance(name, list):
+            for img in name:
+
+                if not isinstance(img, six.string_types):
+                    raise RestException('%s was not a valid string.' % img)
+            else:
```

Comment:
Also maybe it is better to check the sanity of all docker image strings in the list with a for loop raising an error if anything is incorrect. And then delete them in another loop.